File: backend/sdk/contractflow_pro_sdk.py
"""
ContractFlow Pro - Python Client SDK
Comprehensive SDK for interacting with the ContractFlow Pro API
"""

import requests
import json
from typing import Dict, List, Optional, Union, Any
from datetime import datetime
import time


class ContractFlowProSDK:
    """Main SDK class for ContractFlow Pro API interactions"""
    
    def __init__(self, config: Dict[str, Any] = None):
        if config is None:
            config = {}
        
        self.base_url = config.get('base_url', 'http://localhost:5001')
        self.api_base = f"{self.base_url}/api"
        self.token = config.get('token')
        self.timeout = config.get('timeout', 30)
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'ContractFlowPro-Python-SDK/1.0.0'
        })
        
        # Set timeout
        self.session.timeout = self.timeout
    
    def set_token(self, token: str) -> None:
        """Set authentication token"""
        self.token = token
        if token:
            self.session.headers['Authorization'] = f'Bearer {token}'
        else:
            self.session.headers.pop('Authorization', None)
    
    def clear_token(self) -> None:
        """Clear authentication token"""
        self.token = None
        self.session.headers.pop('Authorization', None)
    
    def _make_request(self, method: str, endpoint: str, data: Dict = None, 
                     params: Dict = None) -> Dict[str, Any]:
        """Make HTTP request to API"""
        url = f"{self.api_base}{endpoint}"
        
        try:
            if method.upper() == 'GET':
                response = self.session.get(url, params=params)
            elif method.upper() == 'POST':
                response = self.session.post(url, json=data)
            elif method.upper() == 'PUT':
                response = self.session.put(url, json=data)
            elif method.upper() == 'DELETE':
                response = self.session.delete(url)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            if hasattr(e, 'response') and e.response is not None:
                error_data = e.response.json() if e.response.content else {}
                raise APIError(
                    status_code=e.response.status_code,
                    message=error_data.get('error', str(e)),
                    details=error_data.get('details')
                )
            else:
                raise APIError(status_code=0, message=str(e))
    
    # Health & System Methods
    def get_health(self) -> Dict[str, Any]:
        """Get system health status"""
        return self._make_request('GET', '/health')
    
    def get_api_status(self) -> Dict[str, Any]:
        """Get API status"""
        return self._make_request('GET', '/api/status')
    
    # Authentication Methods
    def login(self, email: str, password: str) -> Dict[str, Any]:
        """User login"""
        data = {'email': email, 'password': password}
        response = self._make_request('POST', '/auth/login', data)
        
        if response.get('success') and response.get('token'):
            self.set_token(response['token'])
        
        return response
    
    def register(self, user_data: Dict[str, Any]) -> Dict[str, Any]:
        """User registration"""
        return self._make_request('POST', '/auth/register', user_data)
    
    def get_profile(self) -> Dict[str, Any]:
        """Get current user profile"""
        return self._make_request('GET', '/auth/profile')
    
    def logout(self) -> Dict[str, Any]:
        """User logout"""
        self.clear_token()
        return {'success': True, 'message': 'Logged out successfully'}
    
    # Contract Management Methods
    def create_ricardian_contract(self, contract_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new Ricardian contract"""
        return self._make_request('POST', '/contracts/ricardian', contract_data)
    
    def get_contract(self, contract_id: str) -> Dict[str, Any]:
        """Get contract by ID"""
        return self._make_request('GET', f'/contracts/{contract_id}')
    
    def list_contracts(self, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """List contracts with optional filters"""
        return self._make_request('GET', '/contracts', params=params)
    
    def update_contract(self, contract_id: str, update_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update contract"""
        return self._make_request('PUT', f'/contracts/{contract_id}', update_data)
    
    def preview_contract(self, contract_data: Dict[str, Any]) -> Dict[str, Any]:
        """Preview contract without authentication"""
        return self._make_request('POST', '/contracts/ricardian/multi-tdp-preview-test', contract_data)
    
    # Dataset Management Methods
    def get_public_datasets(self, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """Get public datasets"""
        return self._make_request('GET', '/datasets/public', params=params)
    
    def search_datasets(self, query: str, filters: Dict[str, Any] = None) -> Dict[str, Any]:
        """Search datasets"""
        params = {'q': query}
        if filters:
            params.update(filters)
        return self._make_request('GET', '/datasets/search', params=params)
    
    def get_tdp_datasets(self, user_id: str) -> Dict[str, Any]:
        """Get TDP user datasets"""
        return self._make_request('GET', f'/tdp/datasets/{user_id}')
    
    def create_dataset(self, dataset_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create new dataset"""
        return self._make_request('POST', '/datasets', dataset_data)
    
    # Infrastructure & Cloud Methods
    def create_training_environment(self, environment_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create training environment"""
        return self._make_request('POST', '/infrastructure/environments', environment_data)
    
    def get_environment(self, environment_id: str) -> Dict[str, Any]:
        """Get environment by ID"""
        return self._make_request('GET', f'/infrastructure/environments/{environment_id}')
    
    def list_environments(self, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """List environments"""
        return self._make_request('GET', '/infrastructure/environments', params=params)
    
    def destroy_environment(self, environment_id: str) -> Dict[str, Any]:
        """Destroy environment"""
        return self._make_request('DELETE', f'/infrastructure/environments/{environment_id}')
    
    def get_cloud_providers(self) -> Dict[str, Any]:
        """Get available cloud providers"""
        return self._make_request('GET', '/infrastructure/cloud-providers')
    
    def get_cloud_provider_details(self, provider: str) -> Dict[str, Any]:
        """Get cloud provider details"""
        return self._make_request('GET', f'/infrastructure/cloud-providers/{provider}')
    
    def estimate_costs(self, estimation_data: Dict[str, Any]) -> Dict[str, Any]:
        """Estimate environment costs"""
        return self._make_request('POST', '/infrastructure/cost-estimation', estimation_data)
    
    # Cloud Credentials Methods
    def store_cloud_credentials(self, credential_data: Dict[str, Any]) -> Dict[str, Any]:
        """Store cloud provider credentials"""
        return self._make_request('POST', '/ccrp/cloud-credentials', credential_data)
    
    def get_cloud_credentials(self, user_id: str) -> Dict[str, Any]:
        """Get stored cloud credentials"""
        return self._make_request('GET', f'/ccrp/cloud-credentials/{user_id}')
    
    def validate_credentials(self, credential_id: str) -> Dict[str, Any]:
        """Validate cloud credentials"""
        return self._make_request('POST', f'/ccrp/cloud-credentials/{credential_id}/validate')
    
    def get_secret_managers(self) -> Dict[str, Any]:
        """Get available secret managers"""
        return self._make_request('GET', '/secret-managers')
    
    # Dashboard Methods
    def get_tdc_dashboard(self, user_id: str) -> Dict[str, Any]:
        """Get TDC dashboard"""
        return self._make_request('GET', f'/tdc/dashboard/{user_id}')
    
    def get_tdp_dashboard(self, user_id: str) -> Dict[str, Any]:
        """Get TDP dashboard"""
        return self._make_request('GET', f'/tdp/dashboard/{user_id}')
    
    def get_ccrp_dashboard(self, user_id: str) -> Dict[str, Any]:
        """Get CCRP dashboard"""
        return self._make_request('GET', f'/ccrp/dashboard/{user_id}')
    
    # Analytics Methods
    def get_contract_analytics(self, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """Get contract analytics"""
        return self._make_request('GET', '/analytics/contracts', params=params)
    
    # User Management Methods
    def list_users(self, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """List users"""
        return self._make_request('GET', '/users', params=params)
    
    def get_user(self, user_id: str) -> Dict[str, Any]:
        """Get user by ID"""
        return self._make_request('GET', f'/users/{user_id}')
    
    # Security & Compliance Methods
    def get_environment_security(self, environment_id: str) -> Dict[str, Any]:
        """Get environment security status"""
        return self._make_request('GET', f'/infrastructure/environments/{environment_id}/security')
    
    def get_compliance_report(self, environment_id: str) -> Dict[str, Any]:
        """Get compliance report"""
        return self._make_request('GET', f'/infrastructure/environments/{environment_id}/compliance')
    
    # Monitoring & Metrics Methods
    def get_environment_metrics(self, environment_id: str, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """Get environment metrics"""
        return self._make_request('GET', f'/infrastructure/environments/{environment_id}/metrics', params=params)
    
    # Utility Methods
    def is_authenticated(self) -> bool:
        """Check if user is authenticated"""
        try:
            self.get_profile()
            return True
        except APIError:
            return False
    
    def batch_create_contracts(self, contracts_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Batch create contracts"""
        results = []
        for data in contracts_data:
            try:
                result = self.create_ricardian_contract(data)
                results.append({'success': True, 'data': result})
            except Exception as e:
                results.append({'success': False, 'error': str(e)})
        return results
    
    def batch_get_contracts(self, contract_ids: List[str]) -> List[Dict[str, Any]]:
        """Batch get contracts"""
        results = []
        for contract_id in contract_ids:
            try:
                result = self.get_contract(contract_id)
                results.append({'success': True, 'data': result})
            except Exception as e:
                results.append({'success': False, 'error': str(e)})
        return results


class ContractManager:
    """Convenience class for contract operations"""
    
    def __init__(self, sdk: ContractFlowProSDK):
        self.sdk = sdk
    
    def create_ai_training_contract(self, dataset_ids: List[str], duration: int, 
                                  terms: str) -> Dict[str, Any]:
        """Create AI training contract with default settings"""
        contract_data = {
            'datasetSelections': [{'datasetId': id, 'individualPrice': 1000} for id in dataset_ids],
            'duration': duration,
            'termsAndConditions': terms,
            'contractType': 'AI_TRAINING',
            'privacyRequirements': {
                'differentialPrivacy': True,
                'maxPrivacyLoss': 0.1
            }
        }
        return self.sdk.create_ricardian_contract(contract_data)
    
    def get_active_contracts(self) -> Dict[str, Any]:
        """Get active contracts"""
        return self.sdk.list_contracts({'status': 'ACTIVE'})
    
    def get_contract_history(self) -> Dict[str, Any]:
        """Get completed contracts"""
        return self.sdk.list_contracts({'status': 'COMPLETED'})


class DatasetManager:
    """Convenience class for dataset operations"""
    
    def __init__(self, sdk: ContractFlowProSDK):
        self.sdk = sdk
    
    def search_by_category(self, category: str, max_price: Optional[float] = None) -> Dict[str, Any]:
        """Search datasets by category"""
        filters = {'category': category}
        if max_price:
            filters['priceMax'] = max_price
        return self.sdk.search_datasets('', filters)
    
    def get_affordable_datasets(self, budget: float) -> Dict[str, Any]:
        """Get datasets within budget"""
        return self.sdk.get_public_datasets({'priceMax': budget})


class InfrastructureManager:
    """Convenience class for infrastructure operations"""
    
    def __init__(self, sdk: ContractFlowProSDK):
        self.sdk = sdk
    
    def provision_azure_environment(self, contract_id: str, vm_size: str = 'Standard_D2s_v3') -> Dict[str, Any]:
        """Provision Azure training environment"""
        environment_data = {
            'contractId': contract_id,
            'cloudProvider': 'AZURE',
            'region': 'eastus',
            'vmSize': vm_size,
            'enableConfidentialComputing': True,
            'enableEncryption': True
        }
        return self.sdk.create_training_environment(environment_data)
    
    def provision_aws_environment(self, contract_id: str, instance_type: str = 't3.medium') -> Dict[str, Any]:
        """Provision AWS training environment"""
        environment_data = {
            'contractId': contract_id,
            'cloudProvider': 'AWS',
            'region': 'us-east-1',
            'vmSize': instance_type,
            'enableConfidentialComputing': True,
            'enableEncryption': True
        }
        return self.sdk.create_training_environment(environment_data)
    
    def monitor_environment(self, environment_id: str) -> Dict[str, Any]:
        """Get environment monitoring data"""
        return self.sdk.get_environment_metrics(environment_id, {'timeRange': '24h'})


class APIError(Exception):
    """Custom exception for API errors"""
    
    def __init__(self, status_code: int, message: str, details: Optional[Dict] = None):
        self.status_code = status_code
        self.message = message
        self.details = details
        super().__init__(self.message)
    
    def __str__(self):
        return f"API Error {self.status_code}: {self.message}"


# Example usage:
"""
from contractflow_pro_sdk import ContractFlowProSDK, ContractManager

# Initialize SDK
sdk = ContractFlowProSDK({
    'base_url': 'https://api.contractflowpro.com',
    'timeout': 30
})

# Login
response = sdk.login('user@example.com', 'password')
print(f"Login successful: {response['success']}")

# Use convenience classes
contract_manager = ContractManager(sdk)
contract = contract_manager.create_ai_training_contract(
    dataset_ids=['DS-001', 'DS-002'],
    duration=30,
    terms='Standard AI training terms'
)

print(f"Contract created: {contract['contract']['contractId']}")
""" 

File: backend/sdk/contractflow-pro-sdk.js
/**
 * ContractFlow Pro - Node.js Client SDK
 * Comprehensive SDK for interacting with the ContractFlow Pro API
 */

const axios = require('axios');

class ContractFlowProSDK {
  constructor(config = {}) {
    this.baseURL = config.baseURL || 'http://localhost:5001';
    this.apiBase = `${this.baseURL}/api`;
    this.token = config.token || null;
    this.timeout = config.timeout || 30000;
    
    // Initialize axios instance
    this.client = axios.create({
      baseURL: this.apiBase,
      timeout: this.timeout,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    // Add request interceptor for authentication
    this.client.interceptors.request.use((config) => {
      if (this.token) {
        config.headers.Authorization = `Bearer ${this.token}`;
      }
      return config;
    });

    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => {
        if (error.response?.status === 401) {
          throw new Error('Authentication failed. Please check your token.');
        }
        throw error;
      }
    );
  }

  /**
   * Set authentication token
   */
  setToken(token) {
    this.token = token;
  }

  /**
   * Clear authentication token
   */
  clearToken() {
    this.token = null;
  }

  /**
   * Health & System
   */
  async getHealth() {
    const response = await this.client.get('/health');
    return response.data;
  }

  async getAPIStatus() {
    const response = await this.client.get('/api/status');
    return response.data;
  }

  /**
   * Authentication
   */
  async login(email, password) {
    const response = await this.client.post('/auth/login', { email, password });
    if (response.data.success && response.data.token) {
      this.setToken(response.data.token);
    }
    return response.data;
  }

  async register(userData) {
    const response = await this.client.post('/auth/register', userData);
    return response.data;
  }

  async getProfile() {
    const response = await this.client.get('/auth/profile');
    return response.data;
  }

  async logout() {
    this.clearToken();
    return { success: true, message: 'Logged out successfully' };
  }

  /**
   * Contract Management
   */
  async createRicardianContract(contractData) {
    const response = await this.client.post('/contracts/ricardian', contractData);
    return response.data;
  }

  async getContract(contractId) {
    const response = await this.client.get(`/contracts/${contractId}`);
    return response.data;
  }

  async listContracts(params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/contracts?${queryParams}`);
    return response.data;
  }

  async updateContract(contractId, updateData) {
    const response = await this.client.put(`/contracts/${contractId}`, updateData);
    return response.data;
  }

  async previewContract(contractData) {
    const response = await this.client.post('/contracts/ricardian/multi-tdp-preview-test', contractData);
    return response.data;
  }

  /**
   * Dataset Management
   */
  async getPublicDatasets(params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/datasets/public?${queryParams}`);
    return response.data;
  }

  async searchDatasets(query, filters = {}) {
    const params = { q: query, ...filters };
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/datasets/search?${queryParams}`);
    return response.data;
  }

  async getTDPDatasets(userId) {
    const response = await this.client.get(`/tdp/datasets/${userId}`);
    return response.data;
  }

  async createDataset(datasetData) {
    const response = await this.client.post('/datasets', datasetData);
    return response.data;
  }

  /**
   * Infrastructure & Cloud
   */
  async createTrainingEnvironment(environmentData) {
    const response = await this.client.post('/infrastructure/environments', environmentData);
    return response.data;
  }

  async getEnvironment(environmentId) {
    const response = await this.client.get(`/infrastructure/environments/${environmentId}`);
    return response.data;
  }

  async listEnvironments(params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/infrastructure/environments?${queryParams}`);
    return response.data;
  }

  async destroyEnvironment(environmentId) {
    const response = await this.client.delete(`/infrastructure/environments/${environmentId}`);
    return response.data;
  }

  async getCloudProviders() {
    const response = await this.client.get('/infrastructure/cloud-providers');
    return response.data;
  }

  async getCloudProviderDetails(provider) {
    const response = await this.client.get(`/infrastructure/cloud-providers/${provider}`);
    return response.data;
  }

  async estimateCosts(estimationData) {
    const response = await this.client.post('/infrastructure/cost-estimation', estimationData);
    return response.data;
  }

  /**
   * Cloud Credentials
   */
  async storeCloudCredentials(credentialData) {
    const response = await this.client.post('/ccrp/cloud-credentials', credentialData);
    return response.data;
  }

  async getCloudCredentials(userId) {
    const response = await this.client.get(`/ccrp/cloud-credentials/${userId}`);
    return response.data;
  }

  async validateCredentials(credentialId) {
    const response = await this.client.post(`/ccrp/cloud-credentials/${credentialId}/validate`);
    return response.data;
  }

  async getSecretManagers() {
    const response = await this.client.get('/secret-managers');
    return response.data;
  }

  /**
   * Dashboards
   */
  async getTDCDashboard(userId) {
    const response = await this.client.get(`/tdc/dashboard/${userId}`);
    return response.data;
  }

  async getTDPDashboard(userId) {
    const response = await this.client.get(`/tdp/dashboard/${userId}`);
    return response.data;
  }

  async getCCRPDashboard(userId) {
    const response = await this.client.get(`/ccrp/dashboard/${userId}`);
    return response.data;
  }

  /**
   * Analytics
   */
  async getContractAnalytics(params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/analytics/contracts?${queryParams}`);
    return response.data;
  }

  /**
   * User Management
   */
  async listUsers(params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/users?${queryParams}`);
    return response.data;
  }

  async getUser(userId) {
    const response = await this.client.get(`/users/${userId}`);
    return response.data;
  }

  /**
   * Security & Compliance
   */
  async getEnvironmentSecurity(environmentId) {
    const response = await this.client.get(`/infrastructure/environments/${environmentId}/security`);
    return response.data;
  }

  async getComplianceReport(environmentId) {
    const response = await this.client.get(`/infrastructure/environments/${environmentId}/compliance`);
    return response.data;
  }

  /**
   * Monitoring & Metrics
   */
  async getEnvironmentMetrics(environmentId, params = {}) {
    const queryParams = new URLSearchParams(params);
    const response = await this.client.get(`/infrastructure/environments/${environmentId}/metrics?${queryParams}`);
    return response.data;
  }

  /**
   * Utility Methods
   */
  async isAuthenticated() {
    try {
      await this.getProfile();
      return true;
    } catch (error) {
      return false;
    }
  }

  async refreshToken() {
    // Implementation depends on refresh token endpoint
    throw new Error('Refresh token not implemented yet');
  }

  /**
   * Batch Operations
   */
  async batchCreateContracts(contractsData) {
    const promises = contractsData.map(data => this.createRicardianContract(data));
    return Promise.allSettled(promises);
  }

  async batchGetContracts(contractIds) {
    const promises = contractIds.map(id => this.getContract(id));
    return Promise.allSettled(promises);
  }

  /**
   * Error Handling
   */
  handleError(error) {
    if (error.response) {
      return {
        status: error.response.status,
        message: error.response.data?.error || error.message,
        details: error.response.data?.details || null
      };
    }
    return {
      status: 0,
      message: error.message,
      details: null
    };
  }
}

/**
 * Convenience Classes for Specific Use Cases
 */
class ContractManager {
  constructor(sdk) {
    this.sdk = sdk;
  }

  async createAITrainingContract(datasetIds, duration, terms) {
    return this.sdk.createRicardianContract({
      datasetSelections: datasetIds.map(id => ({ datasetId: id, individualPrice: 1000 })),
      duration,
      termsAndConditions: terms,
      contractType: 'AI_TRAINING',
      privacyRequirements: {
        differentialPrivacy: true,
        maxPrivacyLoss: 0.1
      }
    });
  }

  async getActiveContracts() {
    return this.sdk.listContracts({ status: 'ACTIVE' });
  }

  async getContractHistory() {
    return this.sdk.listContracts({ status: 'COMPLETED' });
  }
}

class DatasetManager {
  constructor(sdk) {
    this.sdk = sdk;
  }

  async searchByCategory(category, maxPrice = null) {
    const filters = { category };
    if (maxPrice) filters.priceMax = maxPrice;
    return this.sdk.searchDatasets('', filters);
  }

  async getAffordableDatasets(budget) {
    return this.sdk.getPublicDatasets({ priceMax: budget });
  }
}

class InfrastructureManager {
  constructor(sdk) {
    this.sdk = sdk;
  }

  async provisionAzureEnvironment(contractId, vmSize = 'Standard_D2s_v3') {
    return this.sdk.createTrainingEnvironment({
      contractId,
      cloudProvider: 'AZURE',
      region: 'eastus',
      vmSize,
      enableConfidentialComputing: true,
      enableEncryption: true
    });
  }

  async provisionAWSEnvironment(contractId, instanceType = 't3.medium') {
    return this.sdk.createTrainingEnvironment({
      contractId,
      cloudProvider: 'AWS',
      region: 'us-east-1',
      vmSize: instanceType,
      enableConfidentialComputing: true,
      enableEncryption: true
    });
  }

  async monitorEnvironment(environmentId) {
    return this.sdk.getEnvironmentMetrics(environmentId, { timeRange: '24h' });
  }
}

/**
 * Export the SDK and convenience classes
 */
module.exports = {
  ContractFlowProSDK,
  ContractManager,
  DatasetManager,
  InfrastructureManager
};

// Example usage:
/*
const { ContractFlowProSDK, ContractManager } = require('./contractflow-pro-sdk');

const sdk = new ContractFlowProSDK({
  baseURL: 'https://api.contractflowpro.com',
  timeout: 30000
});

// Login
await sdk.login('user@example.com', 'password');

// Use convenience classes
const contractManager = new ContractManager(sdk);
const contract = await contractManager.createAITrainingContract(
  ['DS-001', 'DS-002'],
  30,
  'Standard AI training terms'
);

console.log('Contract created:', contract);
*/ 